backend/src/main/java/ai/verta/modeldb/utils/RdbmsUtils.java

```diff
@@ -270,9 +289,26 @@ public static ObservationEntity generateObservationEntity(
     List<Observation> observationList = new ArrayList<>();
     LOGGER.trace("Converting ObservationEntityListFromObservations");
     if (observationEntityList != null) {
-      for (ObservationEntity observation : observationEntityList) {
-        observationList.add(observation.getProtoObject());
-      }
+      observationList =
+          observationEntityList.stream()
+              .map(
+                  observationEntity -> {
+                    try {
+                      return observationEntity.getProtoObject();
+                    } catch (InvalidProtocolBufferException e) {
+                      throw new RuntimeException(e);
+                    }
+                  })
+              .sorted(
+                  Comparator.comparing(
+                      observation -> {
+                        if (observation.hasArtifact()) {
+                          return observation.getArtifact().getKey();
+                        } else {
+                          return observation.getAttribute().getKey();
+                        }
```

Comment:
Please extract to private method. This will collapse the size of the stream processing block and make the code easier to read.

Reply:
extracted